DependencyInjection/Compiler/EntityListenerPass.php

```diff
@@ -41,35 +45,43 @@ public function process(ContainerBuilder $container)
                     $this->attachToListener($container, $name, $id, $attributes);
                 }
 
-                if (isset($attributes['lazy']) && $attributes['lazy']) {
+                $class = $resolver->getClass();
+
+                if (substr($class, 0, 1) === '%') {
+                    // resolve container parameter first
+                    $class = $container->getParameterBag()->resolveValue($resolver->getClass());
+                }
+                $resolverSupportsLazyListeners = is_a($class, EntityListenerServiceResolver::class, true);
+
+                $lazyByAttribute = isset($attributes['lazy']) && $attributes['lazy'];
+                if ($lazyByAttribute && ! $resolverSupportsLazyListeners) {
+                    throw new InvalidArgumentException(
+                        sprintf('Lazy-loaded entity listeners can only be resolved by a resolver implementing %s.', EntityListenerServiceResolver::class)
```

Comment:
IMO You should break the line in the `sprintf` rather than in the exception, it gives a more even distribution of text:

```php
<?php
throw new InvalidArgumentException(sprintf(
    'Lazy-loaded entity listeners can only be resolved by a resolver implementing %s.',
    EntityListenerServiceResolver::class
));
```